### PR TITLE
implemented tabgroups plugin

### PR DESCRIPTION
**Pull request by serg-kozhemyakin**
@aidanholm can you review this pls? it's same as before just with some lines wrapped

### CODE REVIEW

File: tabgroups/init.lua [hunk 1]
```diff
@@ -0,0 +1,870 @@
...
+            ctime = _gv.tabs[v].ctime,
+            title = v.title,
+            n = i })
+    end
+    if field and order then
+        table.sort(_tmp, function(a, b) return _sort_by_field(field, order, a, b) end)
+    end
+
+    for i, v in ipairs(_tmp) do
+        local title = '*No title*'
```
Comment (aidanholm) [inline tabgroups/init.lua]: `local title = v.title or "*No title*"`; you can also use pango markup instead of asterisks

File: tabgroups/init.lua [hunk 2]
```diff
@@ -0,0 +1,870 @@
...
+        if v.title then
+            title = v.v.title
+        end
+        if v.v == active then
+            table.insert(rows, {"<b>"..((i < 10 and i..' - ') or '')..group_name.."</b>",
```
Comment (aidanholm) [inline tabgroups/init.lua]: table formatting

File: tabgroups/init.lua [hunk 3]
```diff
@@ -0,0 +1,870 @@
...
+                "<b>"..lousy.util.escape(title).."</b>",
+                lousy.util.escape(v.v.uri),
+                _group = group_name, _tab = v.v})
+        else
+            table.insert(rows, {((i < 10 and i..' - ') or '')..lousy.util.escape(group_name),
```
Comment (aidanholm) [inline tabgroups/init.lua]: ditto

File: tabgroups/init.lua [hunk 4]
```diff
@@ -0,0 +1,870 @@
...
+    end
+    return rows
+end
+
+local function _build_tabgroup_menu_grouplist(w, field, order)
+    local rows = {{ "Group name", "Number of tabs", title = true }}
+
+    local _tmp = {}
+    for g, gv in pairs(w2groups[w].groups) do
+        table.insert(_tmp, {name = g,
```
Comment (aidanholm) [inline tabgroups/init.lua]: ditto

File: tabgroups/init.lua [hunk 5]
```diff
@@ -0,0 +1,870 @@
...
+        end
+    end
+    return rows
+end
+
+local function _build_tabgroup_menu_grouplist(w, field, order)
+    local rows = {{ "Group name", "Number of tabs", title = true }}
+
+    local _tmp = {}
+    for g, gv in pairs(w2groups[w].groups) do
```
Comment (aidanholm) [inline tabgroups/init.lua]: this code seems similar to the code in the function above: can it be moved into a helper function?

File: tabgroups/init.lua [hunk 6]
```diff
@@ -0,0 +1,870 @@
...
+        if nb:count() == 0 then
+            w:new_tab(settings.get_setting("window.new_tab_page"), false)
+        end
+
+        -- copy-paste from window.lua, since that handler is only attached to the initial notebook
+        w.view = nil
+        -- Update widgets after tab switch
+        luakit.idle_add(function ()
+            -- Cancel if window already destroyed
+            if not w.win then return end
```
Comment (aidanholm) [inline tabgroups/init.lua]: `not w.win or not w.view`

File: tabgroups/init.lua [hunk 7]
```diff
@@ -0,0 +1,870 @@
...
+    if row and row._group then
+        w:set_mode('tabgroup-menu-rename')
+    end
+end
+
+-- needed for forward declaration
+local show_tabgroup_content, show_tabgroups, move_tab_to_tabgroup_menu
+
+show_tabgroup_content = function (w, tabgroup_name)
+    if tabgroup_name and type(tabgroup_name) == 'table' then
```
Comment (aidanholm) [inline tabgroups/init.lua]: `if type(tabgroup_name) == "table"` works fine: `type(nil)` is `"nil"`

File: tabgroups/init.lua [hunk 8]
```diff
@@ -0,0 +1,870 @@
...
+show_tabgroups = function (w)
+    local rows = build_tabgroup_menu(w)
+    w.menu:build(rows)
+    w.menu:update()
+    local notify = _default_notify ..", ".. "+: show tabs in selected group"
+    w:notify(notify, false)
+end
+
+local function switch_tabgroup_or_tab(w, _, m)
+    local row
```
Comment (aidanholm) [inline tabgroups/init.lua]: `local row = w.menu:get((m and m.count) and m.count+1 or nil)`

File: tabgroups/init.lua [hunk 9]
```diff
@@ -0,0 +1,870 @@
...
+    end
+
+    if expand_group then
+        return _build_tabgroup_menu_grouptabs(w, expand_group, field, order)
+    else
+        return _build_tabgroup_menu_grouplist(w, field, order)
+    end
+end
+
+local _operation = nil;
```
Comment (aidanholm) [inline tabgroups/init.lua]: what if you have multiple operations going on in separate windows?

File: tabgroups/init.lua [hunk 10]
```diff
@@ -0,0 +1,870 @@
...
+    assert(w2groups[w].groups[group]._notebook.type == "notebook")
+    local i = 0
+    local n = w2groups[w].groups[group]._notebook:count()
+    return function()
+        i = i + 1
+        if i <= n then return i, w2groups[w].groups[group]._notebook[i] end
+    end
+end
+
+local function webview2idx(w, group, view)
```
Comment (aidanholm) [inline tabgroups/init.lua]: Can't this just be `view.parent:indexof(view)`? With separate handling if `view.parent ~= w2groups[w].groups[group]._notebook` of course

File: tabgroups/init.lua [hunk 11]
```diff
@@ -0,0 +1,870 @@
...
+        until not w2groups[w].groups[name]
+    else
+        name = _new_tabgroup_prefix .. i
+    end
+    return name
+end
+
+local function grouptabs(w, group)
+    assert(type(group) == 'string')
+    assert(w2groups[w].groups[group])
```
Comment (aidanholm) [inline tabgroups/init.lua]: `group = assert(w2groups[w].groups[group])`, then simplify the rest

File: tabgroups/init.lua [hunk 12]
```diff
@@ -0,0 +1,870 @@
...
+
+            for gn, g in pairs(win_state.tab_groups.groups) do
+                for i, s in ipairs(g.tabs) do
+                    if gn ~= w2groups[w].active then
+                        open_new_tab_in_tabgroup(w, gn, s.uri,  {switch = g.active == i,
+                            session_state = s.session_state,
+                            session_restore = true })
+                    end
+                    if s[i] then
+                        w2groups[w].groups[gn].tabs[i]  = { tab_hit = s[i].tab_hit,
```
Comment (aidanholm) [inline tabgroups/init.lua]: table formatting

File: tabgroups/init.lua [hunk 13]
```diff
@@ -0,0 +1,870 @@
...
+})
+
+local function move_tab_to_tabgroup(w, view, group)
+    local old_group = webview2group(view)
+    if old_group.name ~= group then
+        local tg = create_tabgroup(w, group)
+        local stats = w2groups[w].groups[old_group.name].tabs[view]
+        w2groups[w].groups[old_group.name]._notebook:remove(view)
+        tg._notebook:insert(view)
+        tg.tabs[view] = { tab_hit = stats.tab_hit,
```
Comment (aidanholm) [inline tabgroups/init.lua]: table formatting

File: tabgroups/init.lua [hunk 14]
```diff
@@ -0,0 +1,870 @@
...
+
+settings.register_settings({
+    ["tabgroups.sort_groups_by"] = {
+        type = "string",
+        default = "name asc",
+        validator = function(v)
+            local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
+            if field then
+                field = string.lower(field)
+                if field ~= 'name' and
```
Comment (aidanholm) [inline tabgroups/init.lua]: - `field` and `order` are either both set or both nil
- if they're both nil, that's an error condition: return false
```lua
if not field then return false end
field, order = field:lower(), order:lower()
field = ({name = true, atime = true, mtime = true, ctime = true, hits = true})[field]
order = order == "asc" or order == "desc"
return field and order
```

File: tabgroups/init.lua [hunk 15]
```diff
@@ -0,0 +1,870 @@
...
+            else
+                return false
+            end
+            return true
+        end,
+        desc = [=[ Sort order of groups in tabgroups menu.
+
+Values should be set as 'field order'. Where field is one of next
+
+  name: group name
```
Comment (aidanholm) [inline tabgroups/init.lua]: Use - to add list items, and surround the field name with backticks:
```
...
- `name`: group name
...
```

File: tabgroups/init.lua [hunk 16]
```diff
@@ -0,0 +1,870 @@
...
+                    return false
+                end
+            else
+                return false
+            end
+            return true
+        end,
+        desc = [=[ Sort order of groups in tabgroups menu.
+
+Values should be set as 'field order'. Where field is one of next
```
Comment (aidanholm) [inline tabgroups/init.lua]:
```
Must be in the form "_field_ _order_", where _field_ is one of:
```

File: tabgroups/init.lua [hunk 17]
```diff
@@ -0,0 +1,870 @@
...
+
+  name: group name
+  ctime: time of group creation
+  mtime: time of group modification (updated when we adding or removing new tabs to group)
+  atime: time of group accessing (updated when we switching to this group)
+  hits: number of switching to this group
+
+and order is
```
Comment (aidanholm) [inline tabgroups/init.lua]:
```
and _order_ is one of:
```

File: tabgroups/init.lua [hunk 18]
```diff
@@ -0,0 +1,870 @@
...
+            end
+            return true
+        end,
+        desc = [=[ Sort order of groups in tabgroups menu.
+
+Values should be set as 'field order'. Where field is one of next
+
+  name: group name
+  ctime: time of group creation
+  mtime: time of group modification (updated when we adding or removing new tabs to group)
```
Comment (aidanholm) [inline tabgroups/init.lua]: (adding or removing new tabs) is descriptive enough, I think

File: tabgroups/init.lua [hunk 19]
```diff
@@ -0,0 +1,870 @@
...
+  atime: time of group accessing (updated when we switching to this group)
```
Comment (aidanholm) [inline tabgroups/init.lua]: (switching to this group)

File: tabgroups/init.lua [hunk 20]
```diff
@@ -0,0 +1,870 @@
...
+  hits: number of switching to this group
```
Comment (aidanholm) [inline tabgroups/init.lua]: number of times this group has been switched to

File: tabgroups/init.lua [hunk 21]
```diff
@@ -0,0 +1,870 @@
...
+
+and order is
+
+  asc: sort in ascending order
```
Comment (aidanholm) [inline tabgroups/init.lua]: lists here too

File: tabgroups/init.lua [hunk 22]
```diff
@@ -0,0 +1,870 @@
...
+  desc: sort in descending order
+]=]
+    },
+    ["tabgroups.sort_tabs_by"] = {
+        type = "string",
+        default = "title asc",
```
Comment (aidanholm) [inline tabgroups/init.lua]: same as before

File: tabgroups/init.lua [hunk 23]
```diff
@@ -0,0 +1,870 @@
...
+                    return false
+                end
+            else
+                return false
+            end
+            return true
+        end,
+        desc = [=[ Sort order of tabs in tabgroups menu.
+
+Values should be set as 'field order'. Where field is one of next
```
Comment (aidanholm) [inline tabgroups/init.lua]: same as before

File: tabgroups/init.lua (hunk 9 above)
Comment (serg-kozhemyakin) [reply tabgroups/init.lua]: frankly -- i never opening several different web windows :) preffer one maximized window :) that's why i also preffer tabgroups. but probably you'r right. if different windows shares same lua process then it need to be fixed

File: tabgroups/init.lua (hunk 5 above)
Comment (serg-kozhemyakin) [reply tabgroups/init.lua]: this codes looks similar but dealing with different structures. one with groups another with tabs in group. so i think better leave as is

File: tabgroups/init.lua [hunk 24]
```diff
@@ -0,0 +1,870 @@
...
+end
+
+local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
+    local rows = {{"Group name", "Tab title", "URI", title = true}}
+    local active = current_webview_in_group(w, group_name)
+
+    local _tmp = {}
+    local _gv = w2groups[w].groups[group_name]
+    for i, v in grouptabs(w, group_name) do
+        if not _gv.tabs[v] then
```
[the remark this answers is not in the document]
Comment (serg-kozhemyakin) [reply tabgroups/init.lua]: it's not exactly same. assigning to local tabs will not updated global strucutre.

File: tabgroups/init.lua [hunk 25]
```diff
@@ -0,0 +1,870 @@
...
+
+local function _cleaner()
+    for k, g in pairs(_deleted_groups) do
+        if g and g._notebook then
+            g._notebook:remove_signal("page-removed", _page_removed_cb)
+            g._notebook:remove_signal("page-added", _page_added_cb)
+            g._notebook:remove_signal("switch-page", _switch_page_cb)
+            local i = 0
+            local n = g._notebook:count()
+            while i <= n do
```
[the remark this answers is not in the document]
Comment (serg-kozhemyakin) [reply tabgroups/init.lua]: why for? is it much faster? or something other?

[the remark this answers is not in the document]
Comment (aidanholm) [reply tabgroups/init.lua]: the code should be a lot simpler

File: tabgroups/init.lua (hunk 24 above)
[the remark this answers is not in the document]
Comment (aidanholm) [reply tabgroups/init.lua]:
```lua
local tab = _gv.tabs[v] or { ... }
_gv.tabs[v] = tab
...
```

File: tabgroups/init.lua (hunk 9 above)
Comment (aidanholm) [reply tabgroups/init.lua]: Yep, all the UI code is a single Lua process; the usual way to do this is to have a weak table: `local operation = setmetatable({}, { __mode = "k" })` and then key by `w`.

File: tabgroups/init.lua (hunk 5 above)
Comment (aidanholm) [reply tabgroups/init.lua]: ok no problem then

File: tabgroups/init.lua [hunk 26]
```diff
@@ -0,0 +1,840 @@
...
+local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
+    local rows = {{"Group name", "Tab title", "URI", title = true}}
+    local active = current_webview_in_group(w, group_name)
+
+    local _tmp = {}
+    local _gv = w2groups[w].groups[group_name]
+    for i, v in grouptabs(w, group_name) do
+        local tab = _gv.tabs[v] or { tab_hit = 1, atime = 0, ctime = os.time(), mtime = 0,  }
+        _gv.tabs[v] = tab
+        table.insert(_tmp, {
```
Comment (aidanholm) [inline tabgroups/init.lua]: Table formatting here

File: tabgroups/init.lua [hunk 27]
```diff
@@ -0,0 +1,840 @@
...
+and _order_ is
+
+- `asc`: sort in ascending order
+- `desc`: sort in descending order
+]=]
+    },
+    ["tabgroups.sort_tabs_by"] = {
+        type = "string",
+        default = "title asc",
+        validator = function(v)
```
Comment (aidanholm) [inline tabgroups/init.lua]: This validation function can be made much shorter, the same way as the `tabgroups.sort_groups_by` was.

Comment (serg-kozhemyakin) [reply tabgroups/init.lua]: this is diff for latest changes:
diff --git a/tabgroups/init.lua b/tabgroups/init.lua
index 38a4aec..38ff341 100644
--- a/tabgroups/init.lua
+++ b/tabgroups/init.lua
@@ -300,7 +300,8 @@ session.add_signal("restore", function (state)
                         open_new_tab_in_tabgroup(w, gn, stat.uri,  {
                             switch = g.active == i,
                             session_state = stat.session_state,
-                            session_restore = true })
+                            session_restore = true
+                        })
                     end
                     local group = w2groups[w].groups[gn]
                     group.tabs[i]  = {
@@ -381,7 +382,8 @@ local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
             mtime = tab.mtime,
             ctime = tab.ctime,
             title = v.title,
-            n = i })
+            n = i
+        })
     end
     if field and order then
         table.sort(_tmp, function(a, b) return _sort_by_field(field, order, a, b) end)
@@ -394,13 +396,15 @@ local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
                 "<b>"..((i < 10 and i..' - ') or '')..group_name.."</b>",
                 "<b>"..lousy.util.escape(title).."</b>",
                 lousy.util.escape(v.v.uri),
-                _group = group_name, _tab = v.v})
+                _group = group_name, _tab = v.v
+            })
         else
             table.insert(rows, {
                 ((i < 10 and i..' - ') or '')..lousy.util.escape(group_name),
                 lousy.util.escape(title),
                 lousy.util.escape(v.v.uri),
-                _group = group_name, _tab = v.v})
+                _group = group_name, _tab = v.v
+            })
         end
     end
     return rows
@@ -427,12 +431,14 @@ local function _build_tabgroup_menu_grouplist(w, field, order)
             table.insert(rows, {
                 "<b>"..((i < 10 and i..' - ') or '')..lousy.util.escape(g.name).."</b>",
                 w2groups[w].groups[g.name]._notebook:count(),
-                _group = g.name})
+                _group = g.name
+            })
         else
             table.insert(rows, {
                 ((i < 10 and i..' - ') or '')..lousy.util.escape(g.name),
                 w2groups[w].groups[g.name]._notebook:count(),
-                _group = g.name})
+                _group = g.name
+            })
         end
     end
     return rows
@@ -758,7 +764,7 @@ settings.register_settings({
         default = "name asc",
         validator = function(v)
             local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
-            if not field then return false end
+            if not field or not order then return false end
             field, order = field:lower(), order:lower()
             field = ({name = true, atime = true, mtime = true, ctime = true, hits = true})[field]
             order = order == "asc" or order == "desc"
@@ -785,27 +791,10 @@ and _order_ is
         default = "title asc",
         validator = function(v)
             local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
-            if field then
-                field = string.lower(field)
-                if field ~= 'n' and
-                    field ~= 'title' and
-                    field ~= 'atime' and
-                    field ~= 'mtime' and
-                    field ~= 'ctime' and
-                    field ~= 'hits' then
-                    return false
-                end
-            else
-                return false
-            end
-            if order then
-                order = string.lower(order)
-                if order ~='asc' and order ~= 'desc' then
-                    return false
-                end
-            else
-                return false
-            end
-            return true
+            if not field or not order then return false end
+            field, order = field:lower(), order:lower()
+            field = ({n = true, title = true, atime = true, mtime = true, ctime = true, hits = true})[field]
+            order = order == "asc" or order == "desc"
+            return field and order
         end,
         desc = [=[ Sort order of tabs in tabgroups menu.

File: tabgroups/init.lua (hunk 26 above)
Comment (serg-kozhemyakin) [reply tabgroups/init.lua]: diff from previous version, hope last one :)

```
diff --git a/tabgroups/init.lua b/tabgroups/init.lua
index 38ff341..3c9c2f8 100644
--- a/tabgroups/init.lua
+++ b/tabgroups/init.lua
@@ -127,7 +127,7 @@ end
 local function _page_added_cb(nb, view)
     local g = assert(tablist2group(nb))
     assert(not g.tabs[view])
-    g.tabs[view] = { tab_hit = 0, atime = 0, ctime = os.time(), mtime = 0 }
+    g.tabs[view] = { tab_hit = 0, atime = 0, ctime = os.time(), mtime = 0, }
     g.mtime = os.time();
     view:add_signal("property::uri", function (v)
         local grp = webview2group(v)
@@ -173,7 +173,7 @@ local function create_tabgroup(w, group_name)
             atime = 0,
             mtime = 0,
             ctime = os.time(),
-            tabs = {}
+            tabs = {},
         }

         add_signals_to_notebook(nt)
@@ -218,7 +218,7 @@ local function open_new_tab_in_tabgroup (w, group, uri, opts)
         local pos = tg._notebook:insert((order and order(w, view)) or -1, view)
         assert(tg.tabs[view])
         assert(tg._notebook)
-        webview.set_location(view, { session_state = opts.session_state, uri = uri or opts.uri })
+        webview.set_location(view, { session_state = opts.session_state, uri = uri or opts.uri, })
         if opts.switch then
             tg._notebook:switch(pos)
         end
@@ -246,7 +246,7 @@ end

 window.add_signal("init", function (w)
     local group_name = _get_next_tabgroup_name(w)
-    w2groups[w] = { active = group_name , groups = { }  }
+    w2groups[w] = { active = group_name , groups = {},  }

     local _nb = w.tabs
     local group_nb = widget{type="notebook"}
@@ -261,7 +261,7 @@ window.add_signal("init", function (w)
         atime = os.time(),
         mtime = os.time(),
         ctime = os.time(),
-        tabs = {}
+        tabs = {},
     }

     add_signals_to_notebook(w.tabs)
@@ -300,7 +300,7 @@ session.add_signal("restore", function (state)
                         open_new_tab_in_tabgroup(w, gn, stat.uri,  {
                             switch = g.active == i,
                             session_state = stat.session_state,
-                            session_restore = true
+                            session_restore = true,
                         })
                     end
                     local group = w2groups[w].groups[gn]
@@ -308,7 +308,7 @@ session.add_signal("restore", function (state)
                         tab_hit = stat.tab_hit,
                         atime = stat.atime,
                         ctime = stat.ctime,
-                        mtime = stat.mtime
+                        mtime = stat.mtime,
                     }
                 end
                 local group = w2groups[w].groups[gn]
@@ -324,7 +324,7 @@ end)
 session.add_signal("save", function(state)
     local wins = lousy.util.table.values(window.bywidget)
     for _, w in ipairs(wins) do
-        state[w].tab_groups = { active = w2groups[w].active, groups = {} }
+        state[w].tab_groups = { active = w2groups[w].active, groups = {}, }
         for gn, g in pairs(w2groups[w].groups) do
             local _a = current_webview_in_group(w, gn)
             state[w].tab_groups.groups[gn] = {
@@ -333,7 +333,7 @@ session.add_signal("save", function(state)
                 atime = g.atime,
                 ctime = g.ctime,
                 mtime = g.mtime,
-                tabs = {}
+                tabs = {},
             }
             local dst = state[w].tab_groups.groups[gn]
             for i, v in grouptabs(w, gn) do
@@ -382,7 +382,7 @@ local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
             mtime = tab.mtime,
             ctime = tab.ctime,
             title = v.title,
-            n = i
+            n = i,
         })
     end
     if field and order then
@@ -396,14 +396,16 @@ local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
                 "<b>"..((i < 10 and i..' - ') or '')..group_name.."</b>",
                 "<b>"..lousy.util.escape(title).."</b>",
                 lousy.util.escape(v.v.uri),
-                _group = group_name, _tab = v.v
+                _group = group_name,
+                _tab = v.v,
             })
         else
             table.insert(rows, {
                 ((i < 10 and i..' - ') or '')..lousy.util.escape(group_name),
                 lousy.util.escape(title),
                 lousy.util.escape(v.v.uri),
-                _group = group_name, _tab = v.v
+                _group = group_name,
+                _tab = v.v,
             })
         end
     end
@@ -431,13 +433,13 @@ local function _build_tabgroup_menu_grouplist(w, field, order)
             table.insert(rows, {
                 "<b>"..((i < 10 and i..' - ') or '')..lousy.util.escape(g.name).."</b>",
                 w2groups[w].groups[g.name]._notebook:count(),
-                _group = g.name
+                _group = g.name,
             })
         else
             table.insert(rows, {
                 ((i < 10 and i..' - ') or '')..lousy.util.escape(g.name),
                 w2groups[w].groups[g.name]._notebook:count(),
-                _group = g.name
+                _group = g.name,
             })
         end
     end
@@ -591,7 +593,7 @@ move_tab_to_tabgroup_menu = function (w)
     if number_of_tabgroups(w) > 1 then
         local row = w.menu:get()
         if row and row._group and row._tab then
-            _operation[w] = { op = 'move', _group = row._group, _tab = row._tab }
+            _operation[w] = { op = 'move', _group = row._group, _tab = row._tab, }
             w:set_mode('tabgroup-menu-select')
         end
     end
@@ -624,7 +626,7 @@ new_mode("tabgroup-menu-rename", {
         local groupname = row._group
         w:set_prompt("Enter new name of tabgroup '"..groupname.."' > ")
         w:set_input(groupname)
-        _operation[w] = { op = 'rename', _group = groupname }
+        _operation[w] = { op = 'rename', _group = groupname, }
     end,

     activate = function (w, new_name)
@@ -697,7 +699,7 @@ new_mode("tabgroup-menu-select", {
 })

 add_binds("tabgroup-menu-select",  lousy.util.table.join({
-    { "<Return>", "Select tabgroup.", select_tabgroup }
+    { "<Return>", "Select tabgroup.", select_tabgroup },
 }, menu_binds))

 new_mode("tabgroup-menu", {
```

**Comment by aidanholm** [comment]
Thanks @serg-kozhemyakin, this looks great! Sorry it took so long to get merged :(

**Pull request merged**